src/java.management/share/classes/java/lang/management/ThreadInfo.java

```diff
@@ -544,11 +544,11 @@ public boolean isSuspended() {
     }
 
     /**
-     * Tests if the thread associated with this {@code ThreadInfo}
-     * is executing native code via the Java Native Interface (JNI).
-     * The JNI native code does not include
-     * the virtual machine support code or the compiled native
-     * code generated by the virtual machine.
+     * Tests if the thread associated with this {@code ThreadInfo} is executing
+     * native code. This method returns {@code true} if the thread is executing
+     * a native method or if executing native code invoked using a {@linkplain
+     * java.lang.invoke.MethodHandle method handle} obtained from the
+     * {@linkplain java.lang.foreign.Linker native linker}.
      *
```

Comment:
This area is new to me, but I happened to be in this code few days back. I'm mostly curious on what the actual definition of a thread being in native means.
When a thread is executing any of the following,  does it end up being considered as being in a "native method":

- A syscall (for example, `write()`)
- A C function exposed by a platform specific library
- A JNI method (either part of the JDK or the application) which then may or may not do any syscall or C function call on a platform specific library

Reply:
I would agree, it should state if runtime functions (including those doing a syscall) will be counted here. (For JNi i would not need it to be spelled out, on the other hand it would help, since it makes clear we don’t mean c2 code)

Reply:
This `isInNative` method intends to provide a way to tell if the thread has transitioned from executing Java method to a native method.  JVM interpreter and compiler provide the support for Java language which is why I think such clarification might not be highly necessary.

Reply:
FWIW I would not remove the "does not include ..." information otherwise people may wrongly infer that in the past this method would not return true for VM code or generated code, but that now it does.

Reply:
Note that there is no answer for the "syscall" case mentioned as it depends on the code that makes the syscall: if from a native method then yes; if from the VM then no.

Reply:
> Note that there is no answer for the "syscall" case mentioned as it depends on the code that makes the syscall: if from a native method then yes; if from the VM then no.

Right, if the thread invokes a Java native method then it may or may not invoke a system call, it doesn't matter, the ThreadInfo::isInNative method is just reporting a bit of thread state to say that it is in a native method. Same thing when calling a native function with the FFM API, the native function may be a syscall,  or the native code may invoke a syscall, it doesn't matter to ThreadInfo::isInNative as it just returns a bit of thread state.

Reply:
btw for the following (system/native/runtime/embryotic) Reader thread tseems to be no java method marked "native" on the stack, what case is that?

```
"JDWP Event Helper Thread" #22 daemon prio=10 os_prio=0 cpu=0.00ms elapsed=65.02s tid=0x000001fc4f7e9970 nid=0x32e0 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"JDWP Command Reader" #23 daemon prio=10 os_prio=0 cpu=0.00ms elapsed=65.02s tid=0x000001fc4f7ea480 nid=0x69c4 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Notification Thread" #24 daemon prio=9 os_prio=0 cpu=0.00ms elapsed=64.99s tid=0x000001fc4f7f12a0 nid=0x75f0 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

jshell> mx.getThreadInfo(23)
$17 ==> "JDWP Command Reader" daemon prio=10 Id=23 RUNNABLE (in native)
```
funfact: its not in the list:

```
jshell> mx.getAllThreadIds()
$11 ==> long[8] { 1, 2, 3, 4, 5, 24, 25, 26 }
```

Reply:
Well.. it has to do in so far, that those threads are not hidden - the ThreadInfo sees it and says "isNative". But ok if this is not expected, then indeed it does not need to be documented.

Reply:
> seems to be no java method marked "native" on the stack, what case is that?

A native thread that is attached to the VM but is not currently executing any Java code.

> funfact: its not in the list

JDWP filters its own internal threads out of any thread lists it returns, so this may be a case of using two different API's with different views of the set of threads running.

Reply:
> JDWP filters its own internal threads out of any thread lists it returns, so this may be a case of using two different API's with different views of the set of threads running.

I think the issue that Bernd is running into is a long standing anomaly between ThreadMXBean.getAllThreadIds() and the ThreadMXBean.getThreadInfo methods. The former uses a ThreadsListEnumerator with include_jvmti_agent_threads=false so it filters out the JVMTI agent threads. The latter doesn't do the filtering so if you have threadId of a JVMTI agent thread then you can get its thread info.

Reply:
I've created [JDK-8321095](https://bugs.openjdk.org/browse/JDK-8321095) to track the inconsistency on whether JVMTI agent threads are hidden are not. It seems this inconsistency goes back a long way, to JDK 8 at least.

Reply:
> > funfact: its not in the list
> 
> JDWP filters its own internal threads out of any thread lists it returns, so this may be a case of using two different API's with different views of the set of threads running.

JDWP does filter out its own threads (but not other JVMTI agent threads), but are we talking about JDWP here? It looks like Thread.print dcmd is producing the thread list that includes the JDWP threads, and as pointed out, ThreadMXBean.getAllThreadIds() does not include the JDWP threads (this is due to ThreadMXBean doing the filtering of JVMTI agent threads, not JDWP).

Note that `threadId` is clearly specified by ThreadMXBean, but there is no specification for the Thread.print output, so using unlabeled values from Thread.print as an argument to ThreadMXBean.getThreadInfo() seems somewhat presumptuous.